{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "graphic-attraction",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1m\r\n",
      "============================== Info about spaCy ==============================\u001b[0m\r\n",
      "\r\n",
      "spaCy version    3.1.3                         \r\n",
      "Location         /opt/anaconda3/envs/ontoSpacy/lib/python3.9/site-packages/spacy\r\n",
      "Platform         macOS-10.16-x86_64-i386-64bit \r\n",
      "Python version   3.9.2                         \r\n",
      "Pipelines        en_core_web_sm (3.1.0)        \r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!python -m spacy info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "acknowledged-minority",
   "metadata": {},
   "outputs": [],
   "source": [
    "%reset -f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "vital-architecture",
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# One time run.\n",
    "# ! python -m spacy download en_core_web_sm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "wicked-password",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "distributed-reach",
   "metadata": {},
   "outputs": [],
   "source": [
    "#from pronto import Ontology\n",
    "import spacy\n",
    "from spacy.tokens import Doc, Span, Token\n",
    "from spacy.matcher import PhraseMatcher\n",
    "from spacy.util import filter_spans\n",
    "from spacy.language import Language\n",
    "from spacy.lang.en import English\n",
    "from spacy import displacy\n",
    "import os\n",
    "import pandas as pd\n",
    "import json\n",
    "\n",
    "from dframcy import DframCy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "acknowledged-silver",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11581\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CURIE</th>\n",
       "      <th>name</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEBI:25213</td>\n",
       "      <td>metal cation</td>\n",
       "      <td>biolink:ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEBI:25212</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>biolink:ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEBI:25216</td>\n",
       "      <td>metalloporphyrin</td>\n",
       "      <td>biolink:ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UBERON:0011595</td>\n",
       "      <td>jaw region</td>\n",
       "      <td>biolink:AnatomicalEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEBI:76807</td>\n",
       "      <td>EC 3.5.1.* (non-peptide linear amide C-N hydro...</td>\n",
       "      <td>biolink:ChemicalSubstance</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            CURIE                                               name  \\\n",
       "0     CHEBI:25213                                       metal cation   \n",
       "1     CHEBI:25212                                         metabolite   \n",
       "2     CHEBI:25216                                   metalloporphyrin   \n",
       "3  UBERON:0011595                                         jaw region   \n",
       "4     CHEBI:76807  EC 3.5.1.* (non-peptide linear amide C-N hydro...   \n",
       "\n",
       "                    category  \n",
       "0  biolink:ChemicalSubstance  \n",
       "1  biolink:ChemicalSubstance  \n",
       "2  biolink:ChemicalSubstance  \n",
       "3   biolink:AnatomicalEntity  \n",
       "4  biolink:ChemicalSubstance  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols = ['CUI', 'source', 'CURIE', 'name', 'description', 'category']\n",
    "envoTerms = pd.read_csv(os.path.join('terms', 'envo_termlist.tsv'), sep='\\t', low_memory=False, names=cols)\n",
    "envoTerms = envoTerms.drop(['CUI', 'source','description'], axis=1)\n",
    "envoTerms['CURIE']=envoTerms['CURIE'].str.strip('_SYNONYM')\n",
    "envoTerms = envoTerms.drop_duplicates()\n",
    "print(len(envoTerms))\n",
    "envoTerms.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "instant-sound",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 663 ms, sys: 21.4 ms, total: 685 ms\n",
      "Wall time: 684 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "nlp = English() # spacy.load(\"en_core_web_sm\")\n",
    "# Set terms and patterns\n",
    "terms = {}\n",
    "patterns = []\n",
    "for curie, name, category in envoTerms.to_records(index=False):\n",
    "    if name is not None:\n",
    "        terms[name.lower()] = {'id': curie, 'category': category}\n",
    "        patterns.append(nlp(name))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "unknown-gross",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "### old code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "former-precipitation",
   "metadata": {
    "code_folding": [],
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'class ENVOExtractor(object):\\n    \\n    name = \\'envo_extractor\\'\\n    \\n    def __init__(self, doc, label=\"ENVO\"):\\n        self.label = label\\n        \\n        print(\\'Loading ENVO\\')\\n        envo = Ontology(\"http://purl.obolibrary.org/obo/envo.owl\")\\n        \\n        self.terms = {}\\n        patterns = []\\n        ner_terms = len(envo)\\n        print(\"Importing terms\")\\n        \\n        # iterate over terms in ontology\\n        for idx, term in envo.terms.items():\\n            # if term has a name\\n            if term.name is not None:\\n                self.terms[term.name.lower()] = {\\'id\\': term.id}\\n                patterns.append(nlp(term.name))\\n        \\n        # Initialize matcher & add patterns\\n        self.matcher = PhraseMatcher(nlp.vocab)\\n        self.matcher.add(label, None, *patterns)\\n        \\n        # Set extensions to tokens, spans and docs\\n        Token.set_extension(\\'is_envo_term\\', default=False, force=True)\\n        Token.set_extension(\"envo_id\", default=False, force=True)\\n        Token.set_extension(\"merged_concept\", default=False, force=True)\\n        Doc.set_extension(\"has_envo_ids\", getter=self.has_envo_ids, force=True)\\n        Doc.set_extension(\"envo_ids\", default=[], force=True)\\n        Span.set_extension(\"has_envo_ids\", getter=self.has_envo_ids, force=True)\\n        \\n    def __call__(self, doc):\\n        matches = self.matcher(doc)\\n        spans = [Span(doc, start, end, label=self.label) for matchId, start, end in matches]\\n        doc.ents = spans\\n        \\n        for i, span in enumerate(spans):\\n          span._.set(\"has_envo_ids\", True)\\n          for token in span:\\n               token._.set(\"is_envo_term\", True)\\n               token._.set(\"envo_id\", self.terms[span.text.lower()][\"id\"])\\n\\n        with doc.retokenize() as retokenizer:\\n            for span in filter_spans(spans):\\n                retokenizer.merge(span, attrs={\"_\": {\"merged_concept\": True}})\\n                doc._.envo_ids = list(doc._.envo_ids) + [span]\\n\\n        return doc\\n        \\n    # Setter function for doc level\\n    def has_envo_ids(self, tokens):\\n        return any([t._.get(\"is_envo_term\") for t in tokens])\\n        '"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''class ENVOExtractor(object):\n",
    "    \n",
    "    name = 'envo_extractor'\n",
    "    \n",
    "    def __init__(self, doc, label=\"ENVO\"):\n",
    "        self.label = label\n",
    "        \n",
    "        print('Loading ENVO')\n",
    "        envo = Ontology(\"http://purl.obolibrary.org/obo/envo.owl\")\n",
    "        \n",
    "        self.terms = {}\n",
    "        patterns = []\n",
    "        ner_terms = len(envo)\n",
    "        print(\"Importing terms\")\n",
    "        \n",
    "        # iterate over terms in ontology\n",
    "        for idx, term in envo.terms.items():\n",
    "            # if term has a name\n",
    "            if term.name is not None:\n",
    "                self.terms[term.name.lower()] = {'id': term.id}\n",
    "                patterns.append(nlp(term.name))\n",
    "        \n",
    "        # Initialize matcher & add patterns\n",
    "        self.matcher = PhraseMatcher(nlp.vocab)\n",
    "        self.matcher.add(label, None, *patterns)\n",
    "        \n",
    "        # Set extensions to tokens, spans and docs\n",
    "        Token.set_extension('is_envo_term', default=False, force=True)\n",
    "        Token.set_extension(\"envo_id\", default=False, force=True)\n",
    "        Token.set_extension(\"merged_concept\", default=False, force=True)\n",
    "        Doc.set_extension(\"has_envo_ids\", getter=self.has_envo_ids, force=True)\n",
    "        Doc.set_extension(\"envo_ids\", default=[], force=True)\n",
    "        Span.set_extension(\"has_envo_ids\", getter=self.has_envo_ids, force=True)\n",
    "        \n",
    "    def __call__(self, doc):\n",
    "        matches = self.matcher(doc)\n",
    "        spans = [Span(doc, start, end, label=self.label) for matchId, start, end in matches]\n",
    "        doc.ents = spans\n",
    "        \n",
    "        for i, span in enumerate(spans):\n",
    "          span._.set(\"has_envo_ids\", True)\n",
    "          for token in span:\n",
    "               token._.set(\"is_envo_term\", True)\n",
    "               token._.set(\"envo_id\", self.terms[span.text.lower()][\"id\"])\n",
    "\n",
    "        with doc.retokenize() as retokenizer:\n",
    "            for span in filter_spans(spans):\n",
    "                retokenizer.merge(span, attrs={\"_\": {\"merged_concept\": True}})\n",
    "                doc._.envo_ids = list(doc._.envo_ids) + [span]\n",
    "\n",
    "        return doc\n",
    "        \n",
    "    # Setter function for doc level\n",
    "    def has_envo_ids(self, tokens):\n",
    "        return any([t._.get(\"is_envo_term\") for t in tokens])\n",
    "        '''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "breathing-embassy",
   "metadata": {},
   "source": [
    "### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "better-packing",
   "metadata": {},
   "outputs": [],
   "source": [
    "@Language.component('envo_extractor')\n",
    "def envo_extractor(doc):\n",
    "    \n",
    "    matches = matcher(doc)\n",
    "    \n",
    "    spans = [Span(doc, start, end, label = 'ENVO') for matchId, start, end in matches]\n",
    "    \n",
    "    #doc.ents = spans\n",
    "    doc.spans['ENVO'] = spans\n",
    "    \n",
    "    for i, span in enumerate(spans):\n",
    "        span._.set(\"has_envo_ids\", True)\n",
    "        for token in span:\n",
    "            token._.set(\"is_envo_term\", True)\n",
    "            token._.set(\"envo_id\", terms[span.text.lower()][\"id\"])\n",
    "            token._.set(\"category\", terms[span.text.lower()][\"category\"])\n",
    "\n",
    "    # Add ENVO labelld spans along with built-in ones\n",
    "    doc.ents += tuple(filter_spans(doc.spans['ENVO']))\n",
    "\n",
    "    return doc\n",
    "\n",
    "# Setter function for doc level\n",
    "def has_envo_ids(self, tokens):\n",
    "    return any([t._.get(\"is_envo_term\") for t in tokens])\n",
    "\n",
    "def resolve_substrings(matcher, doc, i, matches):\n",
    "    # Get the current match and create tuple of entity label, start and end.\n",
    "    # Append entity to the doc's entity. (Don't overwrite doc.ents!)\n",
    "    print('###########')\n",
    "    match_id, start, end = matches[i]\n",
    "    entity = Span(doc, start, end, label=\"DUPLICATE\")\n",
    "    doc.ents += (entity,)\n",
    "    #doc.spans[\"ENVO\"] = entity\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "scientific-output",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 379 ms, sys: 28.9 ms, total: 408 ms\n",
      "Wall time: 419 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[('tok2vec', <spacy.pipeline.tok2vec.Tok2Vec at 0x7fe26927cf40>),\n",
       " ('tagger', <spacy.pipeline.tagger.Tagger at 0x7fe238409e50>),\n",
       " ('parser', <spacy.pipeline.dep_parser.DependencyParser at 0x7fe288947d60>),\n",
       " ('attribute_ruler',\n",
       "  <spacy.pipeline.attributeruler.AttributeRuler at 0x7fe269236e00>),\n",
       " ('lemmatizer',\n",
       "  <spacy.lang.en.lemmatizer.EnglishLemmatizer at 0x7fe2692c6880>),\n",
       " ('ner', <spacy.pipeline.ner.EntityRecognizer at 0x7fe288947b80>),\n",
       " ('envo_extractor', <function __main__.envo_extractor(doc)>)]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "nlp = spacy.load(\"en_core_web_sm\")\n",
    "matcher = PhraseMatcher(nlp.vocab)\n",
    "matcher.add(\"ENVO\", None, *patterns, on_match = resolve_substrings)\n",
    "nlp.add_pipe('envo_extractor', after='ner')\n",
    "nlp.pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "occasional-assets",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set extensions to tokens, spans and docs\n",
    "Token.set_extension('is_envo_term', default=False, force=True)\n",
    "Token.set_extension(\"envo_id\", default=False, force=True)\n",
    "#Token.set_extension(\"merged_concept\", default=False, force=True)\n",
    "Token.set_extension(\"category\", default=False, force=True)\n",
    "Doc.set_extension(\"has_envo_ids\", getter=has_envo_ids, force=True)\n",
    "Doc.set_extension(\"envo_ids\", default=[], force=True)\n",
    "Span.set_extension(\"has_envo_ids\", getter=has_envo_ids, force=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "offensive-christmas",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\"></br>However, while \n",
       "<mark class=\"entity\" style=\"background: #7aecec; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Dehalococcoides\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ORG</span>\n",
       "</mark>\n",
       " spp. </br>have received much attention with respect to genome sequencing \n",
       "<mark class=\"entity\" style=\"background: #e4e7d2; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    two\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">CARDINAL</span>\n",
       "</mark>\n",
       " published complete genomes, </br>\n",
       "<mark class=\"entity\" style=\"background: #e4e7d2; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    three\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">CARDINAL</span>\n",
       "</mark>\n",
       " draft genomes , only limited efforts have been undertaken to elucidate genomes of </br>halorespiring \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    bacteria\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       " belonging to the low-GC Gram-positive \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    bacteria\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       ", which comprise one of </br>the largest groups of environmental isolates able to grow on the expense of reductive dehalogenation. </br>In fact, until recently no genome sequencing project has yet been initiated \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    towards\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       " \n",
       "<mark class=\"entity\" style=\"background: #bfeeb7; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Dehalobacter spp.\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">PRODUCT</span>\n",
       "</mark>\n",
       ",</br>an intriguing \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    group\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       " of isolates that are restricted in their metabolism to halorespiration, similar </br>to \n",
       "<mark class=\"entity\" style=\"background: #bfeeb7; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Dehalococcoides\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">PRODUCT</span>\n",
       "</mark>\n",
       " spp. That is why we have now elucidated the genome sequence of Dehalobacter sp. </br>strain E1, which grows in syntrophic coculture with a fermenative Sedimentibacter sp \n",
       "<mark class=\"entity\" style=\"background: #7aecec; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Maphosa\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ORG</span>\n",
       "</mark>\n",
       ", , unpublished . </br>In line with the above mentioned lack of information with respect to their genomic microdiversity, </br>we here propose the genomic sequencing of a \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    number\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       " of closely related strains of the </br>\n",
       "<mark class=\"entity\" style=\"background: #bfeeb7; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Gram\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">PRODUCT</span>\n",
       "</mark>\n",
       "-positive halorespiring phyla \n",
       "<mark class=\"entity\" style=\"background: #7aecec; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Desulfitobacterium\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ORG</span>\n",
       "</mark>\n",
       " and \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Dehalobacter\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">FAC</span>\n",
       "</mark>\n",
       ", which, while highly related </br>phylogenetically, differ significantly in their biodegradative activities and eco-physiological features, </br>such as \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    source\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       " of isolation, spectrum of \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    electron\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       " donors, and alternative \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    electron\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       " acceptors. </br>In addition, we propose to include a \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    number\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       " of populations that did not undergo a lengthy </br>in vitro isolation procedure, which can be expected to impose non-natural selection pressures. </br>This has become feasible because of recent developments in the areas of whole genome amplication </br>of \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    small\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       " amounts of genomic DNA, in combination with high throughput cultivation of environmental</br>microorganisms at the microcolony scale , . The recently obtained \n",
       "<mark class=\"entity\" style=\"background: #aa9cfc; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Fe III\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">PERSON</span>\n",
       "</mark>\n",
       " -reducing ‚</br>\n",
       "<mark class=\"entity\" style=\"background: #aa9cfc; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Deferrispora aromatica\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">PERSON</span>\n",
       "</mark>\n",
       " gen. nov. sp. \n",
       "<mark class=\"entity\" style=\"background: #aa9cfc; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    nov\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">PERSON</span>\n",
       "</mark>\n",
       ".‚Äù \n",
       "<mark class=\"entity\" style=\"background: #feca74; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Kunapuli\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">GPE</span>\n",
       "</mark>\n",
       ", , unpublished results , </br>phylogenetically most closely related to \n",
       "<mark class=\"entity\" style=\"background: #feca74; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Desulfitobacterium\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">GPE</span>\n",
       "</mark>\n",
       " and \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Dehalobacter\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">FAC</span>\n",
       "</mark>\n",
       ", </br>is rather dedicated to degradation of a \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    wide\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       " \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    range\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       " of aromatic hydrocarbons under \n",
       "<mark class=\"entity\" style=\"background: #aa9cfc; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Fe III\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">PERSON</span>\n",
       "</mark>\n",
       " -reducing conditions, </br>and as such the \n",
       "<mark class=\"entity\" style=\"background: #e4e7d2; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    first\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ORDINAL</span>\n",
       "</mark>\n",
       " non-proteobacterial isolate with this feature. </br>The availability of these genome sequences will provide pivotal information to </br>further our understanding of evolution, speciation and \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    functionality\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       " of this environmentally highly</br>relevant \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    group\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       " of \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    bacteria\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       ". The here proposed sequencing project is strongly supported by several </br>researchers leading in the \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    field\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       " of \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    anaerobic\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       " degradation of environmental pollutants. </br>Scientists agreed to actively participate in sequence annotation, \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    building\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       "</br>on their strong expertise from previous and \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    current\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       " genome sequence projects. </br></div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "test = \"\"\"\n",
    "However, while Dehalococcoides spp. \n",
    "have received much attention with respect to genome sequencing two published complete genomes, \n",
    "three draft genomes , only limited efforts have been undertaken to elucidate genomes of \n",
    "halorespiring bacteria belonging to the low-GC Gram-positive bacteria, which comprise one of \n",
    "the largest groups of environmental isolates able to grow on the expense of reductive dehalogenation. \n",
    "In fact, until recently no genome sequencing project has yet been initiated towards Dehalobacter spp.,\n",
    "an intriguing group of isolates that are restricted in their metabolism to halorespiration, similar \n",
    "to Dehalococcoides spp. That is why we have now elucidated the genome sequence of Dehalobacter sp. \n",
    "strain E1, which grows in syntrophic coculture with a fermenative Sedimentibacter sp Maphosa, , unpublished . \n",
    "In line with the above mentioned lack of information with respect to their genomic microdiversity, \n",
    "we here propose the genomic sequencing of a number of closely related strains of the \n",
    "Gram-positive halorespiring phyla Desulfitobacterium and Dehalobacter, which, while highly related \n",
    "phylogenetically, differ significantly in their biodegradative activities and eco-physiological features, \n",
    "such as source of isolation, spectrum of electron donors, and alternative electron acceptors. \n",
    "In addition, we propose to include a number of populations that did not undergo a lengthy \n",
    "in vitro isolation procedure, which can be expected to impose non-natural selection pressures. \n",
    "This has become feasible because of recent developments in the areas of whole genome amplication \n",
    "of small amounts of genomic DNA, in combination with high throughput cultivation of environmental\n",
    "microorganisms at the microcolony scale , . The recently obtained Fe III -reducing ‚\n",
    "Deferrispora aromatica gen. nov. sp. nov.‚Äù Kunapuli, , unpublished results , \n",
    "phylogenetically most closely related to Desulfitobacterium and Dehalobacter, \n",
    "is rather dedicated to degradation of a wide range of aromatic hydrocarbons under Fe III -reducing conditions, \n",
    "and as such the first non-proteobacterial isolate with this feature. \n",
    "The availability of these genome sequences will provide pivotal information to \n",
    "further our understanding of evolution, speciation and functionality of this environmentally highly\n",
    "relevant group of bacteria. The here proposed sequencing project is strongly supported by several \n",
    "researchers leading in the field of anaerobic degradation of environmental pollutants. \n",
    "Scientists agreed to actively participate in sequence annotation, building\n",
    "on their strong expertise from previous and current genome sequence projects. \n",
    "\"\"\"\n",
    "doc = nlp(test)\n",
    "displacy.render(doc, style='ent')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ambient-criticism",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token_text</th>\n",
       "      <th>token_start</th>\n",
       "      <th>token_end</th>\n",
       "      <th>token_pos_</th>\n",
       "      <th>token_tag_</th>\n",
       "      <th>token_dep_</th>\n",
       "      <th>token_head</th>\n",
       "      <th>token_ent_type_</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>bacteria</td>\n",
       "      <td>237</td>\n",
       "      <td>245</td>\n",
       "      <td>NOUN</td>\n",
       "      <td>NNS</td>\n",
       "      <td>pobj</td>\n",
       "      <td>of</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>bacteria</td>\n",
       "      <td>284</td>\n",
       "      <td>292</td>\n",
       "      <td>NOUN</td>\n",
       "      <td>NNS</td>\n",
       "      <td>pobj</td>\n",
       "      <td>to</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>towards</td>\n",
       "      <td>496</td>\n",
       "      <td>503</td>\n",
       "      <td>ADP</td>\n",
       "      <td>IN</td>\n",
       "      <td>prep</td>\n",
       "      <td>initiated</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>group</td>\n",
       "      <td>537</td>\n",
       "      <td>542</td>\n",
       "      <td>NOUN</td>\n",
       "      <td>NN</td>\n",
       "      <td>appos</td>\n",
       "      <td>project</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>number</td>\n",
       "      <td>979</td>\n",
       "      <td>985</td>\n",
       "      <td>NOUN</td>\n",
       "      <td>NN</td>\n",
       "      <td>pobj</td>\n",
       "      <td>of</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>source</td>\n",
       "      <td>1236</td>\n",
       "      <td>1242</td>\n",
       "      <td>NOUN</td>\n",
       "      <td>NN</td>\n",
       "      <td>pobj</td>\n",
       "      <td>as</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>electron</td>\n",
       "      <td>1269</td>\n",
       "      <td>1277</td>\n",
       "      <td>NOUN</td>\n",
       "      <td>NN</td>\n",
       "      <td>compound</td>\n",
       "      <td>donors</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>electron</td>\n",
       "      <td>1302</td>\n",
       "      <td>1310</td>\n",
       "      <td>NOUN</td>\n",
       "      <td>NN</td>\n",
       "      <td>compound</td>\n",
       "      <td>acceptors</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>number</td>\n",
       "      <td>1360</td>\n",
       "      <td>1366</td>\n",
       "      <td>NOUN</td>\n",
       "      <td>NN</td>\n",
       "      <td>dobj</td>\n",
       "      <td>include</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>small</td>\n",
       "      <td>1611</td>\n",
       "      <td>1616</td>\n",
       "      <td>ADJ</td>\n",
       "      <td>JJ</td>\n",
       "      <td>amod</td>\n",
       "      <td>amounts</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344</th>\n",
       "      <td>wide</td>\n",
       "      <td>1990</td>\n",
       "      <td>1994</td>\n",
       "      <td>ADJ</td>\n",
       "      <td>JJ</td>\n",
       "      <td>amod</td>\n",
       "      <td>range</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>range</td>\n",
       "      <td>1995</td>\n",
       "      <td>2000</td>\n",
       "      <td>NOUN</td>\n",
       "      <td>NN</td>\n",
       "      <td>pobj</td>\n",
       "      <td>of</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>390</th>\n",
       "      <td>functionality</td>\n",
       "      <td>2267</td>\n",
       "      <td>2280</td>\n",
       "      <td>NOUN</td>\n",
       "      <td>NN</td>\n",
       "      <td>conj</td>\n",
       "      <td>speciation</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>group</td>\n",
       "      <td>2321</td>\n",
       "      <td>2326</td>\n",
       "      <td>NOUN</td>\n",
       "      <td>NN</td>\n",
       "      <td>pobj</td>\n",
       "      <td>of</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>bacteria</td>\n",
       "      <td>2330</td>\n",
       "      <td>2338</td>\n",
       "      <td>NOUN</td>\n",
       "      <td>NNS</td>\n",
       "      <td>pobj</td>\n",
       "      <td>of</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416</th>\n",
       "      <td>field</td>\n",
       "      <td>2438</td>\n",
       "      <td>2443</td>\n",
       "      <td>NOUN</td>\n",
       "      <td>NN</td>\n",
       "      <td>pobj</td>\n",
       "      <td>in</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>anaerobic</td>\n",
       "      <td>2447</td>\n",
       "      <td>2456</td>\n",
       "      <td>ADJ</td>\n",
       "      <td>JJ</td>\n",
       "      <td>amod</td>\n",
       "      <td>degradation</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>building</td>\n",
       "      <td>2565</td>\n",
       "      <td>2573</td>\n",
       "      <td>VERB</td>\n",
       "      <td>VBG</td>\n",
       "      <td>advcl</td>\n",
       "      <td>participate</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443</th>\n",
       "      <td>current</td>\n",
       "      <td>2618</td>\n",
       "      <td>2625</td>\n",
       "      <td>ADJ</td>\n",
       "      <td>JJ</td>\n",
       "      <td>conj</td>\n",
       "      <td>previous</td>\n",
       "      <td>ENVO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        token_text  token_start  token_end token_pos_ token_tag_ token_dep_  \\\n",
       "39        bacteria          237        245       NOUN        NNS       pobj   \n",
       "49        bacteria          284        292       NOUN        NNS       pobj   \n",
       "86         towards          496        503        ADP         IN       prep   \n",
       "94           group          537        542       NOUN         NN      appos   \n",
       "171         number          979        985       NOUN         NN       pobj   \n",
       "211         source         1236       1242       NOUN         NN       pobj   \n",
       "217       electron         1269       1277       NOUN         NN   compound   \n",
       "222       electron         1302       1310       NOUN         NN   compound   \n",
       "234         number         1360       1366       NOUN         NN       dobj   \n",
       "279          small         1611       1616        ADJ         JJ       amod   \n",
       "344           wide         1990       1994        ADJ         JJ       amod   \n",
       "345          range         1995       2000       NOUN         NN       pobj   \n",
       "390  functionality         2267       2280       NOUN         NN       conj   \n",
       "397          group         2321       2326       NOUN         NN       pobj   \n",
       "399       bacteria         2330       2338       NOUN        NNS       pobj   \n",
       "416          field         2438       2443       NOUN         NN       pobj   \n",
       "418      anaerobic         2447       2456        ADJ         JJ       amod   \n",
       "434       building         2565       2573       VERB        VBG      advcl   \n",
       "443        current         2618       2625        ADJ         JJ       conj   \n",
       "\n",
       "      token_head token_ent_type_  \n",
       "39            of            ENVO  \n",
       "49            to            ENVO  \n",
       "86     initiated            ENVO  \n",
       "94       project            ENVO  \n",
       "171           of            ENVO  \n",
       "211           as            ENVO  \n",
       "217       donors            ENVO  \n",
       "222    acceptors            ENVO  \n",
       "234      include            ENVO  \n",
       "279      amounts            ENVO  \n",
       "344        range            ENVO  \n",
       "345           of            ENVO  \n",
       "390   speciation            ENVO  \n",
       "397           of            ENVO  \n",
       "399           of            ENVO  \n",
       "416           in            ENVO  \n",
       "418  degradation            ENVO  \n",
       "434  participate            ENVO  \n",
       "443     previous            ENVO  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dframcy = DframCy(nlp)\n",
    "annotation_df = dframcy.to_dataframe(doc)\n",
    "annotation_df[annotation_df['token_ent_type_']=='ENVO']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "express-reduction",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "## Gotcha!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "burning-coordinate",
   "metadata": {
    "hidden": true,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              CURIE                name                  category\n",
      "747  UBERON:0003102  surface structures  biolink:AnatomicalEntity\n",
      "750  UBERON:0003102   surface structure  biolink:AnatomicalEntity\n"
     ]
    }
   ],
   "source": [
    "##########\n",
    "# IMPORTANT GOTCHAS!\n",
    "# While creating termlists make sure terms that are the same after lemmatization are assigned the same ID.\n",
    "# This dictates to undo the'_SYNONYM' added for synonyms\n",
    "##########\n",
    "print(envoTerms[envoTerms['name'].str.contains('surface structure')])\n",
    "test1 = \"\"\"surface structures.\"\"\"\n",
    "doc = nlp(test1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "electronic-coral",
   "metadata": {
    "hidden": true,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "               CURIE                  name               category\n",
      "10248  ENVO:02000009  plant tissue culture  biolink:OntologyClass\n",
      "10271  ENVO:02000007        tissue culture  biolink:OntologyClass\n"
     ]
    }
   ],
   "source": [
    "print(envoTerms[envoTerms['name'].str.contains('tissue culture')])\n",
    "test2 = \"\"\"tissue culture\"\"\"\n",
    "doc = nlp(test2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "democratic-sustainability",
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">\n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    tissue culture\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">ENVO</span>\n",
       "</mark>\n",
       "</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "doc\n",
    "displacy.render(doc, style='ent')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
